### PR TITLE
Fixes to Pipe

### PR DESCRIPTION
**Pull request by davidfowl**
- Avoid clearing out the writing head in more cases
- Allow returning a fully consumed block even if writing is in progress if it isn't writing using the current block.
- Synchronize call to Advance

Contributes to #38911

### CODE REVIEW

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 1]
```diff
@@ -497,18 +498,32 @@ private void AdvanceReader(BufferSegment consumedSegment, int consumedIndex, Buf
...
 
-                        // Reset the writing head to null if it's the return block
-                        // then null it out as we're about to reset that memory
-                        if (_writingHead == returnEnd)
+                        returnEnd = nextBlock;
+                    }
+
+                    if (consumedIndex == returnEnd.Length)
+                    {
+                        // If the writing head isn't block we're about to return, then we can move to the next one
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: "isn't block" => "isn't this block"?

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 2]
```diff
@@ -312,18 +313,21 @@ internal bool CommitUnsynchronized()
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal void Advance(int bytes)
         {
-            if ((uint)bytes > (uint)_writingHeadMemory.Length)
+            lock (_sync)
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: You can remove the AggressiveInlining.

Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: cc @benaadams sorry dude, locks mean no more inlining 😆

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: So much lock for writer that doesn't support concurrency 😢

**Pull request merged**